test/utils/image/csi_manifest.go

```diff
@@ -0,0 +1,133 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package image
+
+import (
+	"bytes"
+	"fmt"
+	"io/fs"
+	"regexp"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/kubernetes/test/e2e/testing-manifests"
+)
+
+// All of the image tags are of the format k8s.gcr.io/sig-storage/hostpathplugin:v1.7.3.
+var imageRE = regexp.MustCompile(`^(.*)/([^/:]*):(.*)$`)
+
+// appendCSIImageConfigs extracts image repo, name and version from
+// the YAML files under test/e2e/testing-manifests/storage-csi and
+// creates new config entries  for them.
+func appendCSIImageConfigs(configs map[int]Config) {
+	embeddedFS := testing_manifests.GetE2ETestingManifestsFS().EmbeddedFS
+
+	// We add our images with index numbers that start after the highest existing number.
+	index := 0
+	for i := range configs {
+		if i > index {
+			index = i
+		}
+	}
+
+	err := fs.WalkDir(embeddedFS, "storage-csi", func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() || !strings.HasSuffix(path, ".yaml") {
+			return nil
+		}
+		data, err := embeddedFS.ReadFile(path)
+		if err != nil {
+			return err
+		}
+
+		// Split at the "---" separator before working on
+		// individual item. Only works for .yaml.
+		//
+		// We need to split ourselves because we need access
+		// to each original chunk of data for
+		// runtime.DecodeInto. kubectl has its own
+		// infrastructure for this, but that is a lot of code
+		// with many dependencies.
+		items := bytes.Split(data, []byte("\n---"))
+		for i, item := range items {
+			// We don't care what the actual type is. We just
+			// unmarshal into generic maps and then look up images.
+			var object interface{}
+			if err := yaml.Unmarshal(item, &object); err != nil {
+				return fmt.Errorf("decode item #%d in %s: %v",
+					i, path, err)
+			}
+
+			// Will be called for all image strings.
+			visit := func(value string) {
+				parts := imageRE.FindStringSubmatch(value)
+				if parts == nil {
+					return
+				}
+				config := Config{parts[1], parts[2], parts[3]}
+				for _, otherConfig := range configs {
+					if otherConfig == config {
+						return
+					}
+				}
+				index++
+				configs[index] = config
+			}
+
+			// These paths match plain Pods and more complex types
+			// like Deployments.
+			findStrings(object, visit, "spec", "containers", "image")
+			findStrings(object, visit, "spec", "template", "spec", "containers", "image")
+
+		}
+		return nil
+
+	})
+	if err != nil {
+		panic(err)
+	}
```

Comment:
this can be worrisome but UT will make sure this won't happen(especially when yaml.Unmarshal goes wrong)